### PR TITLE
created streaming api and data collector

### PR DESCRIPTION
**Pull request by ericxtchen**
sprint 2

### CODE REVIEW

File: python_api/stream/client.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        while True:
+            # Read exactly 4-byte header
+            header = _recv_exact(self._sock, _HEADER_BYTES)
+            if header is None:
+                break  # server closed connection gracefully
+
+            (payload_length,) = struct.unpack(_HEADER_FMT, header)
+
+            # Guard against absurdly large frames (e.g. corrupt stream)
+            if payload_length > 10 * 1024 * 1024:  # 10 MB hard cap
```
Comment (ThomasMbrice) [inline python_api/stream/client.py]: 100 * 1024 * 1024

Comment (ThomasMbrice) [inline python_api/stream/client.py]: if payload_length > 100 * 1024 * 1024: # 100 MB hard cap

File: python_api/adapter.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+        self._tick_count = 0
+        self._sim_time   = 0.0
+
+
+    def get_world(self) -> None:
+        return None  # no real world object in mock
+
+    def get_spawn_points(self) -> List[Dict[str, float]]:
+        """Return synthetic spawn point dicts spread on a grid."""
+        import math
```
Comment (ThomasMbrice) [inline python_api/adapter.py]: import math at top of file

File: python_api/adapter.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+        return [
+            {"x": float(i * 2), "y": math.sin(i) * 5.0, "z": 0.0}
```
Comment (ThomasMbrice) [inline python_api/adapter.py]: overflow, underflow?

File: python_api/adapter.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+            for i in range(self._num_spawn_points)
+        ]
+
+    def get_blueprint_library(self) -> "_MockBlueprintLibrary":
+        return _MockBlueprintLibrary()
+
+
+    def make_spawn_command(self, blueprint: Any, spawn_point: Any) -> Dict[str, Any]:
```
Comment (ThomasMbrice) [inline python_api/adapter.py:251]: why do you have 2 implmenations of the same method?

File: python_api/adapter.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        """Return a plain dict that spawn_walkers_batch() can consume."""
+        return {"blueprint": blueprint, "spawn_point": spawn_point}
+
+    def spawn_walker(self, spawn_point: Any, blueprint: Any) -> int:
+        actor_id = self._next_id
+        self._next_id += 1
```
Comment (ThomasMbrice) [inline python_api/adapter.py]: no use a hash, these IDs are not strictly unique

File: python_api/adapter.py [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+            self._next_id += 1
+            self._actors[actor_id] = _make_mock_actor(actor_id)
+            responses.append(_MockBatchResponse(actor_id=actor_id, error=None))
+        return responses
+
+    def get_actor(self, actor_id: int) -> Optional[Dict[str, Any]]:
+        return self._actors.get(actor_id)
+
+    def destroy_actor(self, actor_id: int) -> bool:
+        if actor_id in self._actors:
```
Comment (ThomasMbrice) [inline python_api/adapter.py:275]: if actors are cached later on that will be a massive problem

File: python_api/adapter.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+            "ethical_context": {
+                "visible_actors": []   # filled in by DataCollector
+            },
+            "vehicle_state": {
+                "position": [0.0, 0.0, 0.0],
+                "velocity": [5.0, 0.0, 0.0],
+                "controls": {"throttle": 0.5, "steer": 0.0, "brake": 0.0},
+            },
+            "scene_metadata": {
+                "weather": "clear_noon",
```
Comment (ThomasMbrice) [inline python_api/adapter.py:303]: make sure its documented that these aren't constants, it might be confusing to later modifications why the map wont change

File: python_api/adapter.py [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+
+    def __init__(self, actor_id: int, error: Optional[str]) -> None:
+        self.actor_id = actor_id
+        self.error    = error  # None or empty string = success, non-empty = failure
+
+
+class _MockBlueprintLibrary:
+    """Minimal stand-in for carla.BlueprintLibrary."""
+
+    _BLUEPRINTS = [
```
Comment (ThomasMbrice) [inline python_api/adapter.py]: no

File: python_api/adapter.py [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+        ("walker.pedestrian.0001", "female"),
+        ("walker.pedestrian.0002", "male"),
+        ("walker.pedestrian.0003", "female"),
+        ("walker.pedestrian.0004", "male"),
+        ("walker.pedestrian.0005", "female"),
+        ("walker.pedestrian.0006", "male"),
+    ]
```
Comment (ThomasMbrice) [inline python_api/adapter.py]: if youre going to do sm like this make it at least a set of attributes and pick randomly

File: python_api/adapter.py [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+
+    def __init__(self, blueprint_id: str, gender: str) -> None:
+        self.id          = blueprint_id
+        self._attributes = {"gender": gender}
+
+    def get_attribute(self, key: str) -> "_MockAttribute":
+        return _MockAttribute(self._attributes.get(key, ""))
+
+
+class _MockAttribute:
```
Comment (ThomasMbrice) [inline python_api/adapter.py:368]: honestly you dont even really need mock, its fine that its here just make sure it optional and does load on default

File: python_api/collector.py [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+            RuntimeError: If the collector is already running.
+        """
+        if hz <= 0:
+            raise ValueError(f"hz must be > 0, got {hz}")
+        if self._running:
+            raise RuntimeError("DataCollector is already running. Call stop() first.")
+
+        self._interval = 1.0 / hz
+        self._running  = True
+        self._thread   = threading.Thread(
```
Comment (ThomasMbrice) [inline python_api/collector.py:106]: it might be useful to track pid

File: test/test_stream.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+            conn, _ = server_sock.accept()
+            # Header claiming 20 MB payload
+            conn.sendall(struct.pack('>I', 20 * 1024 * 1024))
+            conn.close()
+            server_sock.close()
+
+        t = threading.Thread(target=send_oversized, daemon=True)
+        t.start()
+
+        client = StreamClient(host='127.0.0.1', port=BASE_PORT + 42, timeout=3.0)
```
Comment (ThomasMbrice) [inline test/test_stream.py:265]: overall these tests are good, they dont really attack edge cases thought that might show up in actuul deployment like maybe if there is a timeout and the client saves info but the server resumes connection, assuming that the port number doesnt change now the client is reciveing duplicate objects

**Review by ThomasMbrice — CHANGES_REQUESTED** [review]
